docs/spec/staking/transactions.md

```diff
@@ -70,214 +64,280 @@ If either the `Description` (excluding `DateBonded` which is constant),
-    candidate = loadCandidate(store, tx.PubKey)
-    if candidate == nil return
-	return delegateWithCandidate(tx, candidate)
-
-delegateWithCandidate(tx TxDelegate, candidate Candidate):
-    if candidate.Status == Revoked return
-
-    if candidate.Status == Bonded 
-	    poolAccount = params.HoldBonded
-    else 
-	    poolAccount = params.HoldUnbonded
-	
-    err = transfer(sender, poolAccount, tx.Amount)
-    if err != nil return 
+    pool = getPool()
+    if validator.Status == Revoked return
 
-    bond = loadDelegatorBond(store, sender, tx.PubKey)
-    if bond == nil then bond = DelegatorBond(tx.PubKey, rational.Zero, Coin(0), Coin(0))
-	
-    issuedDelegatorShares = addTokens(tx.Amount, candidate)
-    bond.Shares += issuedDelegatorShares
-	
-    saveCandidate(store, candidate)
-    saveDelegatorBond(store, sender, bond)
-    saveGlobalState(store, gs)
-    return 
-
-addTokens(amount coin.Coin, candidate Candidate):
-    if candidate.Status == Bonded 
-	    gs.BondedPool += amount
-	    issuedShares = amount / exchangeRate(gs.BondedShares, gs.BondedPool)
-	    gs.BondedShares += issuedShares
-    else 
-	    gs.UnbondedPool += amount
-	    issuedShares = amount / exchangeRate(gs.UnbondedShares, gs.UnbondedPool)
-	    gs.UnbondedShares += issuedShares
+    delegation = getDelegatorBond(DelegatorAddr, ValidatorAddr)
+    if delegation == nil then delegation = NewDelegation(DelegatorAddr, ValidatorAddr)
 	
-    candidate.GlobalStakeShares += issuedShares
+    validator, pool, issuedDelegatorShares = validator.addTokensFromDel(tx.Amount, pool)
+    delegation.Shares += issuedDelegatorShares
     
-    if candidate.IssuedDelegatorShares.IsZero() 
-        exRate = rational.One
-    else
-        exRate = candidate.GlobalStakeShares / candidate.IssuedDelegatorShares
-	
-    issuedDelegatorShares = issuedShares / exRate
-    candidate.IssuedDelegatorShares += issuedDelegatorShares
-    return issuedDelegatorShares
-	
-exchangeRate(shares rational.Rat, tokenAmount int64):
-    if shares.IsZero() then return rational.One
-    return tokenAmount / shares
-    	
+    setDelegation(delegation)
+    updateValidator(validator)
+    setPool(pool)
+    return 
 ```
 
-### TxUnbond 
+### TxStartUnbonding
 
 Delegator unbonding is defined with the following transaction:
 
 ```golang
-type TxUnbond struct {
-    PubKey crypto.PubKey
-    Shares rational.Rat 
+type TxStartUnbonding struct {
+	DelegatorAddr sdk.Address 
+	ValidatorAddr sdk.Address 
+	Shares        string      
 }
 
-unbond(tx TxUnbond):    
-    bond = loadDelegatorBond(store, sender, tx.PubKey)
-    if bond == nil return 
-    if bond.Shares < tx.Shares return 
-	
-    bond.Shares -= tx.Shares
-
-    candidate = loadCandidate(store, tx.PubKey)
-	
-    revokeCandidacy = false
-    if bond.Shares.IsZero() 
-	    if sender == candidate.Owner and candidate.Status != Revoked then revokeCandidacy = true then removeDelegatorBond(store, sender, bond)
+startUnbonding(tx TxStartUnbonding):    
```

Comment:
This looks like Go code, not pseudocode.

We can have pseudocode that looks like Go, but IMO going as far as `!bond.Shares.GT(sdk.ZeroRat())` is very confusing, we should just say `! (bond.shares > 0)` - also, not clear from reading this why we can't just check `bond.shares == 0`...